### PR TITLE
Feat (hadamard): support region expansion

### PR DESCRIPTION
**Pull request by Giuseppe5**
## Risk Highlight

- [ ] This PR includes code from another work (please detail).
- [ ] This PR contains API-breaking changes.
- [ ] This PR depends on work in another PR (please provide links/details).
- [ ] This PR introduces new dependencies (please detail).
- [ ] There are coverage gaps not covered by tests.
- [ ] Documentation updates required in subsequent PR.

## Checklist

- [ ] Code comments added to any hard-to-understand areas, if applicable.
- [ ] Changes generate no new warnings.
- [ ] Updated any relevant tests, if applicable.
- [ ] No conflicts with destination `dev` branch.
- [ ] I reviewed my own code changes.
- [ ] Initial CI/CD passing.
- [ ] 1+ reviews given, and any review issues addressed and approved.
- [ ] Post-review full CI/CD passing.

### CODE REVIEW

File: src/brevitas/graph/equalize.py [hunk 1]
```diff
@@ -1519,7 +1555,25 @@ def find_module(self, model, regions: List, prefix=''):
...
                 full_name = prefix + '.' + name if prefix != '' else name
-                self.find_module(module, regions, full_name)
+                self.find_module(module, regions, full_name, blacklist_layers)
+
+    def find_module_by_name(self, model: nn.Module, regions: List[Region], prefix: str = ''):
+        """
+        Iterate through the model looking at immediate children of every module to look for named modules.
+        This allows us to stop the search when we meet a top-level module that is supported.
+        """
+        if prefix in self.layers_to_expand:
```
Comment (Giuseppe5) [inline src/brevitas/graph/equalize.py:1565]: Implicit assumption is that we use this function to layers_to_expand.

Fair assumption as we use the other function (`find_module`) for the supported layerwise modules.
It can also become a new args for the function I guess, but not much value in that at the moment

File: src/brevitas/graph/equalize.py [hunk 2]
```diff
@@ -155,6 +156,7 @@ class Region:
     sinks: Dict = field(default_factory=dict)
     acts: Tuple = field(default_factory=tuple)
     name_to_module: Dict = field(default_factory=dict)
+    expand_region: bool = False
```
Comment (Giuseppe5) [inline src/brevitas/graph/equalize.py:159]: Not sure if I like putting this here.
The class is used for other things and we use this flag in a very specific way.

I guess this class was also created to account for such issues, so why not? Maybe in the future this flag will be used more

File: src/brevitas/graph/equalize.py [hunk 3]
```diff
@@ -1342,7 +1344,10 @@ def _apply_rotate(
         else:
             try:
                 # Build hadamard rotation matrix
-                rot_mat, K = get_hadK(hidden_dim)
+                original_rot_mat, original_k = rot_mat, K = get_hadK(hidden_dim)
```
Comment (Giuseppe5) [inline src/brevitas/graph/equalize.py:1347]: There could be a better way to do this. In general, we generate this "too early", so the idea is to generate both

File: src/brevitas/graph/equalize.py [hunk 4]
```diff
@@ -1384,27 +1389,44 @@ def _apply_rotate(
 
         for name, indexes in region.sinks.items():
             module = region.get_module_from_name(name)
+            weight_axis = _get_input_axis(module)
+
             # Only "weight" is rotated
-            tensor_names_axis = [("weight", _get_input_axis(module))]
+            if region.expand_region:
+                assert isinstance(module, nn.Linear), "Currently only Linear layers support expanded hadamard"
```
Comment (Giuseppe5) [inline src/brevitas/graph/equalize.py:1396]: The code just below can support any layer type, the issue is more on the activation side since we need to define also the dimension along which the input needs padding. It is not complicated to do tbh, but not necessary atm

File: src/brevitas/graph/equalize.py (hunk 2 above)
Comment (nickfraser) [reply src/brevitas/graph/equalize.py:159]: Could this also be used in the channel-splitting case?

File: src/brevitas/graph/equalize.py [hunk 5]
```diff
@@ -1501,16 +1523,30 @@ def _merge_ln(layer_norm, next_module, scale_bias_by_weight):
...
+            self.layers_to_expand = []
+        self.supported_sinks = ()
 
-    def find_module(self, model, regions: List, prefix=''):
+    def find_module(
+            self,
+            model: nn.Module,
+            regions: List[Region],
+            prefix: str = '',
+            blacklist_layers: Optional[List[str]] = None):
```
Comment (Giuseppe5) [inline src/brevitas/graph/equalize.py:1543]: The exclusion list needs to be passed since we want to exclude all the layers that might be expanded

File: src/brevitas/graph/hadamard.py [hunk 6]
```diff
@@ -73,6 +73,27 @@ def get_hadK(n, transpose=False):
     return hadK, K
 
 
+def find_closest_hadamard_number(orig, n):
+    import math
+
+    values_to_check = [172, 156, 140, 108, 60, 52, 40, 36, 28, 20, 12]
```
Comment (Giuseppe5) [inline src/brevitas/graph/hadamard.py:79]: Hardcoded values as the ones below. Not pretty, not ugly

File: src/brevitas/nn/equalized_layer.py [hunk 7]
```diff
@@ -53,24 +54,40 @@ def forward(self, *args, **kwargs):
...
         super().__init__()
         if had_mat is not None:
             self.had_mat = torch.nn.Parameter(had_mat).cpu()
         else:
             self.had_mat = None
         self.layer = layer
         self.k = k
+        self.expand = expand
+
+    def input_pad(self, inp):
```
Comment (Giuseppe5) [inline src/brevitas/nn/equalized_layer.py:67]: Not very general. Not pretty, not ugly

File: src/brevitas/graph/equalize.py [hunk 8]
```diff
@@ -1342,7 +1344,10 @@ def _apply_rotate(
         else:
             try:
                 # Build hadamard rotation matrix
-                rot_mat, K = get_hadK(hidden_dim)
+                original_rot_mat, original_k = rot_mat, K = get_hadK(hidden_dim)
+
+                hidden_dim = find_closest_hadamard_number(hidden_dim, hidden_dim + 1).cpu().item()
```
Comment (nickfraser) [inline src/brevitas/graph/equalize.py:1349]: `+ 1` hard coded? Shouldn't this be a parameter? E.g., what if I want to expand by _more_ than the next highest Hadamard?

File: src/brevitas/graph/equalize.py (hunk 4 above)
Comment (nickfraser) [inline src/brevitas/graph/equalize.py:1396]: Save convs for a new PR?

File: src/brevitas_examples/llm/main.py [hunk 9]
```diff
@@ -994,6 +1002,14 @@ def parse_args(args, override_defaults={}):
...
             args += [f"--{key}", str(override_defaults[key])]
             del override_defaults[key]
+=======
+    parser.add_argument(
+        '--rotation-layers-to-expand',
+        type=str,
+        default=[],
+        nargs='*',
+        help='A list of module names to expand with hadamard rotation. Default: %(default)s')
+>>>>>>> Feat (hadamard): support region expansion
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/main.py:1012]: Ops

File: src/brevitas/graph/equalize.py [hunk 10]
```diff
@@ -1384,27 +1389,44 @@ def _apply_rotate(
...
         for name, indexes in region.sinks.items():
             module = region.get_module_from_name(name)
+            weight_axis = _get_input_axis(module)
+
             # Only "weight" is rotated
-            tensor_names_axis = [("weight", _get_input_axis(module))]
+            if region.expand_region:
+                assert isinstance(module, nn.Linear), "Currently only Linear layers support expanded hadamard"
+                hidden_dim = module.weight.shape[1]
+                new_hidden = find_closest_hadamard_number(hidden_dim, hidden_dim + 1)
```
Comment (nickfraser) [inline src/brevitas/graph/equalize.py:1398]: `+1` here? Maybe the previous comment should live here?

File: src/brevitas/nn/equalized_layer.py [hunk 11]
```diff
@@ -53,24 +54,40 @@ def forward(self, *args, **kwargs):
...
+        # TODO: This only works for Linear layers. We have an assert in equalize.py to check for this
+        hidden_dim = inp.shape[-1]
+        new_hidden = find_closest_hadamard_number(hidden_dim, hidden_dim + 1)
```
Comment (nickfraser) [inline src/brevitas/nn/equalized_layer.py:70]: Similar here - the `+1` hardcoding...

File: src/brevitas/nn/equalized_layer.py (hunk 7 above)
Comment (nickfraser) [inline src/brevitas/nn/equalized_layer.py:67]: Can this share logic with the block in the `Region` class?

File: src/brevitas_examples/llm/main.py [hunk 12]
```diff
@@ -981,6 +988,7 @@ def parse_args(args, override_defaults={}):
         type=str,
         nargs='*',
         help='A list of tasks for zero_shot evaluation. Default: %(default)s')
+<<<<<<< HEAD
```
Comment (nickfraser) [inline src/brevitas_examples/llm/main.py]: Merge conflict not resolved

File: src/brevitas_examples/llm/main.py [hunk 13]
```diff
@@ -994,6 +1002,14 @@ def parse_args(args, override_defaults={}):
         for key in extra_args_keys:
             args += [f"--{key}", str(override_defaults[key])]
             del override_defaults[key]
+=======
```
Comment (nickfraser) [inline src/brevitas_examples/llm/main.py]: Merge conflict not resolved

**Review by nickfraser — CHANGES_REQUESTED** [review]

File: src/brevitas/graph/equalize.py (hunk 8 above)
Comment (Giuseppe5) [reply src/brevitas/graph/equalize.py:1349]: You would need to do a "+1" expansion twice. We can change the function to accept a "step" argument, to decide how many steps to move forward.
Even then, not sure I want to expose this new `step` argument right now

File: src/brevitas/graph/equalize.py (hunk 2 above)
Comment (Giuseppe5) [reply src/brevitas/graph/equalize.py:159]: Yes and no. When doing channel splitting you basically assume that all the regions are going to be expanded anyway.

File: src/brevitas/nn/equalized_layer.py (hunk 7 above)
Comment (Giuseppe5) [reply src/brevitas/nn/equalized_layer.py:67]: We should, I'll look into that

File: src/brevitas/graph/equalize.py (hunk 4 above)
Comment (Giuseppe5) [reply src/brevitas/graph/equalize.py:1396]: I think so, yeah. I'll open an issue to keep track of that as good first issue

File: src/brevitas/utils/logging.py [hunk 14]
```diff
@@ -12,6 +13,25 @@
 from brevitas.utils.quant_utils import *
 
 
+def setup_logger(name):
```
Comment (nickfraser) [inline src/brevitas/utils/logging.py:16]: :O!

File: src/brevitas/graph/equalize.py [hunk 15]
```diff
@@ -1629,17 +1680,29 @@ def apply(self,
...
-        self.find_module(graph_model, orphan_regions)
+
+        blacklist_orphan_layers = self.blacklist_layers + self.layers_to_expand
+        self.find_module(graph_model, orphan_regions, blacklist_layers=blacklist_orphan_layers)
+
+        if len(expanded_regions) > 0:
+            parameter_number_pre = 0
+            for m in graph_model.parameters():
+                parameter_number_pre += m.numel()
+            logging.info(f"{len(expanded_regions)} layers will be expanded during rotation")
```
Comment (nickfraser) [inline src/brevitas/graph/equalize.py:1695]: Personally I think all logging should be protected with a `verbose` flag. Sometimes we will want to turn these things off (for some specific experiments)

**Review by nickfraser — APPROVED** [review]
Only one note about verbosity flags. I guess this can also be achieved by setting the log level higher globally.

File: src/brevitas/utils/logging.py [hunk 16]
```diff
@@ -12,6 +13,25 @@
...
+
+    logger = logging.getLogger(name)
+    logger.setLevel(logging.DEBUG)
```
Comment (nickfraser) [inline src/brevitas/utils/logging.py:19]: Maybe have the logging level as an optional parameter? You might want to turn off all 'INFO' messages at some point.

File: src/brevitas/graph/equalize.py (hunk 15 above)
Comment (Giuseppe5) [reply src/brevitas/graph/equalize.py:1695]: I will go the pytorch way and have an env variable to set the level.
Default will be "INFO" and user can select DEBUG with BREVITAS_LOGGING=DEBUG

File: src/brevitas/utils/logging.py (hunk 16 above)
Comment (Giuseppe5) [reply src/brevitas/utils/logging.py:19]: See the comment above. I think this will change a lot as we figure out what to do/how to do it, I want to have something reasonable to start and I believe your comment/my proposal above should accomplish just that

File: src/brevitas/config.py [hunk 17]
```diff
@@ -20,6 +20,7 @@ def env_to_bool(name, default):
 JIT_ENABLED = env_to_bool('BREVITAS_JIT', False) and _enabled
 NATIVE_STE_BACKEND_ENABLED = env_to_bool('BREVITAS_NATIVE_STE_BACKEND', False)
 VERBOSE = env_to_bool('BREVITAS_VERBOSE', False)
+LOGGING_LEVEL = os.environ.get('BREVITAS_LOGGING', 'CRITICAL')
```
Comment (nickfraser) [inline src/brevitas/config.py:23]: Love it! <3

**Review by nickfraser — APPROVED** [review]
LGTM!
